File: apps/api/internal/calculation/travel_allowance.go
package calculation

import (
	"fmt"
	"time"

	"github.com/shopspring/decimal"
)

// LocalTravelInput holds parameters for local travel allowance calculation.
type LocalTravelInput struct {
	DistanceKm      decimal.Decimal
	DurationMinutes int
	Rules           []LocalTravelRuleInput
}

// LocalTravelRuleInput represents a single local travel rule for matching.
type LocalTravelRuleInput struct {
	MinDistanceKm      decimal.Decimal
	MaxDistanceKm      *decimal.Decimal
	MinDurationMinutes int
	MaxDurationMinutes *int
	TaxFreeAmount      decimal.Decimal
	TaxableAmount      decimal.Decimal
}

// LocalTravelOutput contains the results of local travel allowance calculation.
type LocalTravelOutput struct {
	Matched        bool
	TaxFreeTotal   decimal.Decimal
	TaxableTotal   decimal.Decimal
	TotalAllowance decimal.Decimal
	MatchedRuleIdx int // index of the matched rule, -1 if none
}

// CalculateLocalTravelAllowance finds the first matching rule by distance/duration and returns the amounts.
// Rules are expected to be pre-sorted by the repository (sort_order ASC, min_distance_km ASC).
// The first rule where the distance and duration fall within the rule's range wins.
func CalculateLocalTravelAllowance(input LocalTravelInput) LocalTravelOutput {
	output := LocalTravelOutput{
		MatchedRuleIdx: -1,
		TaxFreeTotal:   decimal.Zero,
		TaxableTotal:   decimal.Zero,
		TotalAllowance: decimal.Zero,
	}

	for i, rule := range input.Rules {
		// Check distance range
		if input.DistanceKm.LessThan(rule.MinDistanceKm) {
			continue
		}
		if rule.MaxDistanceKm != nil && input.DistanceKm.GreaterThan(*rule.MaxDistanceKm) {
			continue
		}

		// Check duration range
		if input.DurationMinutes < rule.MinDurationMinutes {
			continue
		}
		if rule.MaxDurationMinutes != nil && input.DurationMinutes > *rule.MaxDurationMinutes {
			continue
		}

		// First matching rule wins
		output.Matched = true
		output.MatchedRuleIdx = i
		output.TaxFreeTotal = rule.TaxFreeAmount
		output.TaxableTotal = rule.TaxableAmount
		output.TotalAllowance = rule.TaxFreeAmount.Add(rule.TaxableAmount)
		return output
	}

	return output
}

// ExtendedTravelInput holds parameters for extended travel allowance calculation.
type ExtendedTravelInput struct {
	StartDate        time.Time
	EndDate          time.Time
	ThreeMonthActive bool
	Rule             ExtendedTravelRuleInput
}

// ExtendedTravelRuleInput represents an extended travel rule's rate configuration.
type ExtendedTravelRuleInput struct {
	ArrivalDayTaxFree      decimal.Decimal
	ArrivalDayTaxable      decimal.Decimal
	DepartureDayTaxFree    decimal.Decimal
	DepartureDayTaxable    decimal.Decimal
	IntermediateDayTaxFree decimal.Decimal
	IntermediateDayTaxable decimal.Decimal
	ThreeMonthEnabled      bool
	ThreeMonthTaxFree      decimal.Decimal
	ThreeMonthTaxable      decimal.Decimal
}

// ExtendedTravelOutput contains the results of extended travel allowance calculation.
type ExtendedTravelOutput struct {
	TotalDays        int
	ArrivalDays      int
	DepartureDays    int
	IntermediateDays int
	TaxFreeTotal     decimal.Decimal
	TaxableTotal     decimal.Decimal
	TotalAllowance   decimal.Decimal
	Breakdown        []ExtendedTravelBreakdownItem
}

// ExtendedTravelBreakdownItem describes a single line in the extended travel breakdown.
type ExtendedTravelBreakdownItem struct {
	Description     string
	Days            int
	TaxFreeAmount   decimal.Decimal
	TaxableAmount   decimal.Decimal
	TaxFreeSubtotal decimal.Decimal
	TaxableSubtotal decimal.Decimal
}

// CalculateExtendedTravelAllowance computes the allowance for a multi-day trip.
//
// Day calculation (inclusive):
//   - Same day (1 day): 1 arrival day only
//   - 2 days: 1 arrival day + 1 departure day
//   - 3+ days: 1 arrival day + (N-2) intermediate days + 1 departure day
//
// Three-month rule: if ThreeMonthActive && Rule.ThreeMonthEnabled, intermediate days
// use the reduced three-month rates instead of regular intermediate rates.
func CalculateExtendedTravelAllowance(input ExtendedTravelInput) ExtendedTravelOutput {
	output := ExtendedTravelOutput{
		TaxFreeTotal:   decimal.Zero,
		TaxableTotal:   decimal.Zero,
		TotalAllowance: decimal.Zero,
		Breakdown:      make([]ExtendedTravelBreakdownItem, 0),
	}

	// Calculate total days (inclusive)
	totalDays := int(input.EndDate.Sub(input.StartDate).Hours()/24) + 1
	if totalDays < 1 {
		totalDays = 1
	}
	output.TotalDays = totalDays

	rule := input.Rule

	switch {
	case totalDays == 1:
		// Same day: treat as 1 arrival day
		output.ArrivalDays = 1
		output.DepartureDays = 0
		output.IntermediateDays = 0

		arrivalTaxFree := rule.ArrivalDayTaxFree
		arrivalTaxable := rule.ArrivalDayTaxable

		output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
			Description:     "Arrival day",
			Days:            1,
			TaxFreeAmount:   arrivalTaxFree,
			TaxableAmount:   arrivalTaxable,
			TaxFreeSubtotal: arrivalTaxFree,
			TaxableSubtotal: arrivalTaxable,
		})

		output.TaxFreeTotal = arrivalTaxFree
		output.TaxableTotal = arrivalTaxable

	case totalDays == 2:
		// 2 days: arrival + departure
		output.ArrivalDays = 1
		output.DepartureDays = 1
		output.IntermediateDays = 0

		output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
			Description:     "Arrival day",
			Days:            1,
			TaxFreeAmount:   rule.ArrivalDayTaxFree,
			TaxableAmount:   rule.ArrivalDayTaxable,
			TaxFreeSubtotal: rule.ArrivalDayTaxFree,
			TaxableSubtotal: rule.ArrivalDayTaxable,
		})
		output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
			Description:     "Departure day",
			Days:            1,
			TaxFreeAmount:   rule.DepartureDayTaxFree,
			TaxableAmount:   rule.DepartureDayTaxable,
			TaxFreeSubtotal: rule.DepartureDayTaxFree,
			TaxableSubtotal: rule.DepartureDayTaxable,
		})

		output.TaxFreeTotal = rule.ArrivalDayTaxFree.Add(rule.DepartureDayTaxFree)
		output.TaxableTotal = rule.ArrivalDayTaxable.Add(rule.DepartureDayTaxable)

	default:
		// 3+ days: arrival + intermediate + departure
		intermediateDays := totalDays - 2
		output.ArrivalDays = 1
		output.DepartureDays = 1
		output.IntermediateDays = intermediateDays

		// Arrival day
		output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
			Description:     "Arrival day",
			Days:            1,
			TaxFreeAmount:   rule.ArrivalDayTaxFree,
			TaxableAmount:   rule.ArrivalDayTaxable,
			TaxFreeSubtotal: rule.ArrivalDayTaxFree,
			TaxableSubtotal: rule.ArrivalDayTaxable,
		})

		// Intermediate days
		intDays := decimal.NewFromInt(int64(intermediateDays))
		var intTaxFree, intTaxable decimal.Decimal

		if input.ThreeMonthActive && rule.ThreeMonthEnabled {
			// Use reduced three-month rates
			intTaxFree = rule.ThreeMonthTaxFree
			intTaxable = rule.ThreeMonthTaxable
			output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
				Description:     fmt.Sprintf("Intermediate days (three-month rule) x%d", intermediateDays),
				Days:            intermediateDays,
				TaxFreeAmount:   intTaxFree,
				TaxableAmount:   intTaxable,
				TaxFreeSubtotal: intTaxFree.Mul(intDays),
				TaxableSubtotal: intTaxable.Mul(intDays),
			})
		} else {
			intTaxFree = rule.IntermediateDayTaxFree
			intTaxable = rule.IntermediateDayTaxable
			output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
				Description:     fmt.Sprintf("Intermediate days x%d", intermediateDays),
				Days:            intermediateDays,
				TaxFreeAmount:   intTaxFree,
				TaxableAmount:   intTaxable,
				TaxFreeSubtotal: intTaxFree.Mul(intDays),
				TaxableSubtotal: intTaxable.Mul(intDays),
			})
		}

		// Departure day
		output.Breakdown = append(output.Breakdown, ExtendedTravelBreakdownItem{
			Description:     "Departure day",
			Days:            1,
			TaxFreeAmount:   rule.DepartureDayTaxFree,
			TaxableAmount:   rule.DepartureDayTaxable,
			TaxFreeSubtotal: rule.DepartureDayTaxFree,
			TaxableSubtotal: rule.DepartureDayTaxable,
		})

		output.TaxFreeTotal = rule.ArrivalDayTaxFree.
			Add(intTaxFree.Mul(intDays)).
			Add(rule.DepartureDayTaxFree)
		output.TaxableTotal = rule.ArrivalDayTaxable.
			Add(intTaxable.Mul(intDays)).
			Add(rule.DepartureDayTaxable)
	}

	output.TotalAllowance = output.TaxFreeTotal.Add(output.TaxableTotal)

	return output
}

File: src/lib/calculation/travel-allowance.ts
/**
 * Travel Allowance Calculation Utility
 *
 * Pure functions for computing local and extended travel allowances.
 * No Prisma or tRPC dependencies -- suitable for server and client use.
 *
 * Ported from Go: apps/api/internal/calculation/travel_allowance.go
 */

// --- Types ---

export interface LocalTravelRuleInput {
  minDistanceKm: number
  maxDistanceKm: number | null
  minDurationMinutes: number
  maxDurationMinutes: number | null
  taxFreeAmount: number
  taxableAmount: number
}

export interface LocalTravelInput {
  distanceKm: number
  durationMinutes: number
  rules: LocalTravelRuleInput[]
}

export interface LocalTravelOutput {
  matched: boolean
  taxFreeTotal: number
  taxableTotal: number
  totalAllowance: number
  matchedRuleIdx: number
}

export interface ExtendedTravelRuleInput {
  arrivalDayTaxFree: number
  arrivalDayTaxable: number
  departureDayTaxFree: number
  departureDayTaxable: number
  intermediateDayTaxFree: number
  intermediateDayTaxable: number
  threeMonthEnabled: boolean
  threeMonthTaxFree: number
  threeMonthTaxable: number
}

export interface ExtendedTravelInput {
  startDate: Date
  endDate: Date
  threeMonthActive: boolean
  rule: ExtendedTravelRuleInput
}

export interface ExtendedTravelBreakdownItem {
  description: string
  days: number
  taxFreeAmount: number
  taxableAmount: number
  taxFreeSubtotal: number
  taxableSubtotal: number
}

export interface ExtendedTravelOutput {
  totalDays: number
  arrivalDays: number
  departureDays: number
  intermediateDays: number
  taxFreeTotal: number
  taxableTotal: number
  totalAllowance: number
  breakdown: ExtendedTravelBreakdownItem[]
}

// --- Functions ---

/**
 * Finds the first matching local travel rule by distance/duration
 * and returns the amounts.
 *
 * Rules are expected to be pre-sorted by the caller
 * (sort_order ASC, min_distance_km ASC).
 * The first rule where the distance and duration fall within
 * the rule's range wins.
 */
export function calculateLocalTravelAllowance(
  input: LocalTravelInput
): LocalTravelOutput {
  const output: LocalTravelOutput = {
    matched: false,
    taxFreeTotal: 0,
    taxableTotal: 0,
    totalAllowance: 0,
    matchedRuleIdx: -1,
  }

  for (let i = 0; i < input.rules.length; i++) {
    const rule = input.rules[i]!

    // Check distance range
    if (input.distanceKm < rule.minDistanceKm) {
      continue
    }
    if (rule.maxDistanceKm !== null && input.distanceKm > rule.maxDistanceKm) {
      continue
    }

    // Check duration range
    if (input.durationMinutes < rule.minDurationMinutes) {
      continue
    }
    if (
      rule.maxDurationMinutes !== null &&
      input.durationMinutes > rule.maxDurationMinutes
    ) {
      continue
    }

    // First matching rule wins
    output.matched = true
    output.matchedRuleIdx = i
    output.taxFreeTotal = rule.taxFreeAmount
    output.taxableTotal = rule.taxableAmount
    output.totalAllowance = rule.taxFreeAmount + rule.taxableAmount
    return output
  }

  return output
}

/**
 * Computes the allowance for a multi-day trip.
 *
 * Day calculation (inclusive):
 *   - Same day (1 day): 1 arrival day only
 *   - 2 days: 1 arrival day + 1 departure day
 *   - 3+ days: 1 arrival day + (N-2) intermediate days + 1 departure day
 *
 * Three-month rule: if threeMonthActive && rule.threeMonthEnabled,
 * intermediate days use the reduced three-month rates instead of
 * regular intermediate rates.
 */
export function calculateExtendedTravelAllowance(
  input: ExtendedTravelInput
): ExtendedTravelOutput {
  const output: ExtendedTravelOutput = {
    totalDays: 0,
    arrivalDays: 0,
    departureDays: 0,
    intermediateDays: 0,
    taxFreeTotal: 0,
    taxableTotal: 0,
    totalAllowance: 0,
    breakdown: [],
  }

  // Calculate total days (inclusive)
  const msPerDay = 24 * 60 * 60 * 1000
  let totalDays =
    Math.floor(
      (input.endDate.getTime() - input.startDate.getTime()) / msPerDay
    ) + 1
  if (totalDays < 1) {
    totalDays = 1
  }
  output.totalDays = totalDays

  const rule = input.rule

  if (totalDays === 1) {
    // Same day: treat as 1 arrival day
    output.arrivalDays = 1
    output.departureDays = 0
    output.intermediateDays = 0

    output.breakdown.push({
      description: "Arrival day",
      days: 1,
      taxFreeAmount: rule.arrivalDayTaxFree,
      taxableAmount: rule.arrivalDayTaxable,
      taxFreeSubtotal: rule.arrivalDayTaxFree,
      taxableSubtotal: rule.arrivalDayTaxable,
    })

    output.taxFreeTotal = rule.arrivalDayTaxFree
    output.taxableTotal = rule.arrivalDayTaxable
  } else if (totalDays === 2) {
    // 2 days: arrival + departure
    output.arrivalDays = 1
    output.departureDays = 1
    output.intermediateDays = 0

    output.breakdown.push({
      description: "Arrival day",
      days: 1,
      taxFreeAmount: rule.arrivalDayTaxFree,
      taxableAmount: rule.arrivalDayTaxable,
      taxFreeSubtotal: rule.arrivalDayTaxFree,
      taxableSubtotal: rule.arrivalDayTaxable,
    })
    output.breakdown.push({
      description: "Departure day",
      days: 1,
      taxFreeAmount: rule.departureDayTaxFree,
      taxableAmount: rule.departureDayTaxable,
      taxFreeSubtotal: rule.departureDayTaxFree,
      taxableSubtotal: rule.departureDayTaxable,
    })

    output.taxFreeTotal = rule.arrivalDayTaxFree + rule.departureDayTaxFree
    output.taxableTotal = rule.arrivalDayTaxable + rule.departureDayTaxable
  } else {
    // 3+ days: arrival + intermediate + departure
    const intermediateDays = totalDays - 2
    output.arrivalDays = 1
    output.departureDays = 1
    output.intermediateDays = intermediateDays

    // Arrival day
    output.breakdown.push({
      description: "Arrival day",
      days: 1,
      taxFreeAmount: rule.arrivalDayTaxFree,
      taxableAmount: rule.arrivalDayTaxable,
      taxFreeSubtotal: rule.arrivalDayTaxFree,
      taxableSubtotal: rule.arrivalDayTaxable,
    })

    // Intermediate days
    let intTaxFree: number
    let intTaxable: number

    if (input.threeMonthActive && rule.threeMonthEnabled) {
      // Use reduced three-month rates
      intTaxFree = rule.threeMonthTaxFree
      intTaxable = rule.threeMonthTaxable
      output.breakdown.push({
        description: `Intermediate days (three-month rule) x${intermediateDays}`,
        days: intermediateDays,
        taxFreeAmount: intTaxFree,
        taxableAmount: intTaxable,
        taxFreeSubtotal: intTaxFree * intermediateDays,
        taxableSubtotal: intTaxable * intermediateDays,
      })
    } else {
      intTaxFree = rule.intermediateDayTaxFree
      intTaxable = rule.intermediateDayTaxable
      output.breakdown.push({
        description: `Intermediate days x${intermediateDays}`,
        days: intermediateDays,
        taxFreeAmount: intTaxFree,
        taxableAmount: intTaxable,
        taxFreeSubtotal: intTaxFree * intermediateDays,
        taxableSubtotal: intTaxable * intermediateDays,
      })
    }

    // Departure day
    output.breakdown.push({
      description: "Departure day",
      days: 1,
      taxFreeAmount: rule.departureDayTaxFree,
      taxableAmount: rule.departureDayTaxable,
      taxFreeSubtotal: rule.departureDayTaxFree,
      taxableSubtotal: rule.departureDayTaxable,
    })

    output.taxFreeTotal =
      rule.arrivalDayTaxFree +
      intTaxFree * intermediateDays +
      rule.departureDayTaxFree
    output.taxableTotal =
      rule.arrivalDayTaxable +
      intTaxable * intermediateDays +
      rule.departureDayTaxable
  }

  output.totalAllowance = output.taxFreeTotal + output.taxableTotal

  return output
}
